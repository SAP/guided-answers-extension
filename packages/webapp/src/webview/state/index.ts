export { actions, store } from './store';
export { getInitialState } from './reducers';
