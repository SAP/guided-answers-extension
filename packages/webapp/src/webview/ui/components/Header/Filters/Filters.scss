.dialog-filter .ms-Dialog-content {
    overflow: hidden;
}

ul.filter-list {
    list-style-type: none;
    padding-left: 0px;
    overflow: auto;
    height: 100%;
}

.filter-button {
    width: 18px;
    height: 18px;
    border-radius: 2px;
    margin-left: 8px;
}

.filter-button-selected {
    background-color: var(--vscode-button-background);

    &:hover {
        background-color: var(--vscode-button-hoverBackground);
    }
}

.filter-button-selected path {
    fill: var(--vscode-button-foreground);
}
