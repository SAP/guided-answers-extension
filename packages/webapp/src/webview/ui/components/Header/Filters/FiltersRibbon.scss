.clear-filters {
    height: 25px;
    color: var(--vscode-foreground);
    padding: 0 5px 2px 3px;
    text-align: start;
    cursor: pointer;
    background: none;
    border: 0;
    margin-left: 7px;
    &__content {
        vertical-align: middle;
        display: inline-block;
        margin-bottom: 3px;
    }
    svg {
        width: 16px;
        height: 16px;
        margin-right: -2px;
    }
}
.clear-filters:hover {
    background: var(--vscode-menubar-selectionBackground);
    outline: 1px solid var(--vscode-contrastActiveBorder);
    border-radius: 3px;
    > span {
        text-decoration: none;
    }
}
.clear-filters:focus-visible {
    outline: 1px solid var(--vscode-focusBorder, #007fd4);
}

.text-underline {
    text-decoration: underline;
    text-underline-offset: 1px;
}
