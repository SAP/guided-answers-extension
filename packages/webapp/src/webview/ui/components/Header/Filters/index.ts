export * from './Filters';
export * from './FiltersRibbon';
