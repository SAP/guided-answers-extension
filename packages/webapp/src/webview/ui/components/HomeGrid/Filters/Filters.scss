.guided-answer__filter {
    >button {
        padding: 15px;
        width: 100%;
        display: flex;
        gap: 10px;
        flex-flow: row wrap;
        color: var(--vscode-foreground);
        background: var(--vscode-editorWidget-background);
        text-align: left;
        font-weight: 400;
        font-size: 13px;
        cursor: pointer;
        border: none;
        border-radius: 3px;

        &:hover {
            outline: 1px solid var(--vscode-focusBorder, #007fd4);
            outline-offset: -1px;
        }

        &:focus-visible {
            outline: 1px solid var(--vscode-focusBorder, #007fd4);
            outline-offset: -2px;
        }

        strong {
            font-weight: 700;
        }
    }

}
