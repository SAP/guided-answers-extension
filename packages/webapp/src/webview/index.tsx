import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
