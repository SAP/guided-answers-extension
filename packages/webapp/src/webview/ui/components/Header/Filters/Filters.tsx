import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import { AppState } from '../../../../types';
import { actions } from '../../../../state';
import { UiIcons } from '../../UIComponentsLib/Icons';
import { UIIconButton, UIDialog, UICheckbox, UITextInput } from '@sap-ux/ui-components';
import type { ProductFilter, ComponentFilter } from '@sap/guided-answers-extension-types';
import { Stack, IStackTokens } from '@fluentui/react';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';

import './Filters.scss';

/**
 * Sorts products array.
 *
 * @param filters A products array
 * @returns A sorted list of products
 */
export const sortProductFilters = (filters: ProductFilter[]) => {
    return filters.sort((a: ProductFilter, b: ProductFilter) => {
        return a.PRODUCT.localeCompare(b.PRODUCT, undefined, {
            numeric: true,
            sensitivity: 'base'
        });
    });
};

export const sortComponentFilters = (filters: ComponentFilter[]) => {
    type WithNoHyphens = {
        NOHYPHENS?: string;
    };
    type ComponentFilterWithNoHyphens = ComponentFilter & WithNoHyphens;

    const sorting = (a: ComponentFilterWithNoHyphens, b: ComponentFilterWithNoHyphens): any => {
        if (a.NOHYPHENS !== undefined && b.NOHYPHENS !== undefined) {
            return a.NOHYPHENS.localeCompare(b.NOHYPHENS, undefined, {
                numeric: true,
                sensitivity: 'base'
            });
        }
        return undefined;
    };

    return filters
        .map((f: ComponentFilter) => {
            let filter: ComponentFilterWithNoHyphens = f;
            filter.NOHYPHENS = f.COMPONENT.replace(/-/g, '');
            return filter;
        })
        .sort(sorting)
        .map((c: ComponentFilterWithNoHyphens): ComponentFilter => {
            delete c.NOHYPHENS;
            return c;
        });
};

/**
 *
 * @returns Filters for the header
 */
export function Filters() {
    const appState = useSelector<AppState, AppState>((state) => state);
    const PRODUCTS = 'Products';
    const COMPONENTS = 'Components';
    const [isDialogVisible, setDialogVisible] = useState(false);
    const [filter, setFilterType] = useState(PRODUCTS);
    const [productFilters, setProductFilters] = useState(appState.guidedAnswerTreeSearchResult.productFilters);
    const [componentFilters, setComponentFilters] = useState(appState.guidedAnswerTreeSearchResult.componentFilters);
    const [query, setQuery] = useState('');
    const [selectedProductFilters, setSelectedProductFilters] = useState([...appState.selectedProductFilters]);
    const [selectedComponentFilters, setSelectedComponentFilters] = useState([...appState.selectedProductFilters]);
    const isFilterProducts = filter === PRODUCTS;
    const main = {
        ['.ms-TextField-wrapper > .ms-Label']: {
            margin: 0
        },
        height: '400px'
    };

    const verticalGapStackTokens: IStackTokens = {
        childrenGap: 10,
        padding: '10px 0'
    };

    const toggleVisibility = (type: string): void => {
        if (type === PRODUCTS) {
            setFilterType(PRODUCTS);
        } else {
            setFilterType(COMPONENTS);
        }
        setQuery('');
        setDialogVisible(!isDialogVisible);
    };

    const applyFilters = (type: string): void => {
        toggleVisibility(type);
        if (type === PRODUCTS) {
            actions.setProductFilters(selectedProductFilters);
            actions.searchTree({
                query: appState.query,
                filters: {
                    product: selectedProductFilters,
                    component: selectedComponentFilters
                }
            });
        } else {
            actions.setComponentFilters(selectedComponentFilters);
            actions.searchTree({
                query: appState.query,
                filters: {
                    product: selectedProductFilters,
                    component: selectedComponentFilters
                }
            });
        }
    };

    const cancel = () => {
        setSelectedComponentFilters([...appState.selectedComponentFilters]);
        setSelectedProductFilters([...appState.selectedProductFilters]);
        setDialogVisible(false);
    };

    const toggleFilters = (type: string): void => {
        toggleVisibility(type);
        if (type === PRODUCTS) {
            setSelectedProductFilters(
                appState.guidedAnswerTreeSearchResult.productFilters
                    .filter((v) => selectedProductFilters.includes(v.PRODUCT))
                    .map((v) => v.PRODUCT)
            );
            setProductFilters(sortProductFilters(appState.guidedAnswerTreeSearchResult.productFilters));
        } else {
            setSelectedComponentFilters(
                appState.guidedAnswerTreeSearchResult.componentFilters
                    .filter((v) => selectedComponentFilters.includes(v.COMPONENT))
                    .map((v) => v.COMPONENT)
            );
            setComponentFilters(sortComponentFilters(appState.guidedAnswerTreeSearchResult.componentFilters));
        }
    };

    const resetFilter = (): void => {
        setQuery('');
        actions.searchTree({
            query: appState.query,
            filters: {
                product: appState.selectedProductFilters.length > 0 ? appState.selectedProductFilters : [],
                component: appState.selectedComponentFilters.length > 0 ? appState.selectedComponentFilters : []
            }
        });
        setDialogVisible(!isDialogVisible);
    };

    const onChange = (filter: string) => (ev?: any, checked?: boolean) => {
        if (isFilterProducts) {
            if (checked) {
                setSelectedProductFilters([...selectedProductFilters, filter]);
            } else {
                setSelectedProductFilters(selectedProductFilters.filter((e) => e !== filter));
            }
        } else if (checked) {
            setSelectedComponentFilters([...selectedComponentFilters, filter]);
        } else {
            setSelectedComponentFilters(selectedComponentFilters.filter((e) => e !== filter));
        }
    };

    useEffect(() => {
        setSelectedProductFilters(appState.selectedProductFilters);
        setSelectedComponentFilters(appState.selectedComponentFilters);
    }, [appState.selectedProductFilters, appState.selectedComponentFilters]);

    const filterType: any = {
        Products: {
            title: 'Filter Products',
            visibility: isDialogVisible,
            apply: () => applyFilters(PRODUCTS),
            listItems: productFilters.map((productFilter: { PRODUCT: string }) => (
                <UICheckbox
                    key={`${productFilter.PRODUCT}`}
                    label={productFilter.PRODUCT}
                    checked={selectedProductFilters.includes(productFilter.PRODUCT)}
                    onChange={onChange(productFilter.PRODUCT)}
                />
            ))
        },
        Components: {
            title: 'Filter Components',
            visibility: isDialogVisible,
            apply: () => applyFilters(COMPONENTS),
            listItems: componentFilters.map((componentFilter: { COMPONENT: string }) => (
                <UICheckbox
                    key={`${componentFilter.COMPONENT}`}
                    label={componentFilter.COMPONENT}
                    checked={selectedComponentFilters.includes(componentFilter.COMPONENT)}
                    onChange={onChange(componentFilter.COMPONENT)}
                />
            ))
        }
    };

    const searchFilter = (e: any) => {
        const query = e.target.value;

        if (query.length >= 1) {
            setQuery(query);
            let filters: (ProductFilter | ComponentFilter)[] =
                appState.guidedAnswerTreeSearchResult[isFilterProducts ? 'productFilters' : 'componentFilters'];

            filters = filters.filter((f: ProductFilter | ComponentFilter) =>
                isFilterProducts
                    ? (f as ProductFilter).PRODUCT.toLowerCase().includes(query.toLowerCase())
                    : (f as ComponentFilter).COMPONENT.toLowerCase().includes(query.toLowerCase())
            );

            if (isFilterProducts) {
                setProductFilters(sortProductFilters(filters as ProductFilter[]));
            } else {
                setComponentFilters(sortComponentFilters(filters as ComponentFilter[]));
            }
        } else {
            setQuery('');
            setProductFilters(sortProductFilters(appState.guidedAnswerTreeSearchResult.productFilters));
            setComponentFilters(sortComponentFilters(appState.guidedAnswerTreeSearchResult.componentFilters));
        }
    };

    return (
        <>
            <div id="filters">
                <FocusZone
                    direction={FocusZoneDirection.horizontal}
                    isCircularNavigation={true}
                    style={{ display: 'flex' }}>
                    <UIIconButton
                        id="filter-products"
                        iconProps={{ iconName: UiIcons.Table }}
                        onClick={() => toggleFilters(PRODUCTS)}
                        disabled={appState.guidedAnswerTreeSearchResult.productFilters.length === 0}
                        className={`filter-button ${selectedProductFilters.length > 0 ? 'filter-button-selected' : ''}`}
                        primary
                        title="Filter Products"></UIIconButton>
                    <UIIconButton
                        id="filter-components"
                        iconProps={{ iconName: UiIcons.IdTag }}
                        onClick={() => toggleFilters(COMPONENTS)}
                        disabled={appState.guidedAnswerTreeSearchResult.componentFilters.length === 0}
                        primary
                        title="Filter Components"
                        className={`filter-button ${
                            selectedComponentFilters.length > 0 ? 'filter-button-selected' : ''
                        }`}></UIIconButton>
                </FocusZone>
                <UIDialog
                    className="dialog-filter"
                    dialogContentProps={{ title: filterType[filter].title }}
                    isOpen={filterType[filter].visibility}
                    isBlocking={true}
                    acceptButtonText={'Apply Filter'}
                    cancelButtonText={'Cancel'}
                    styles={{ main }}
                    onAccept={() => filterType[filter].apply()}
                    onCancel={cancel}
                    onDismiss={resetFilter}>
                    <UITextInput placeholder="Search" value={query} onChange={searchFilter} />
                    <FocusZone
                        direction={FocusZoneDirection.domOrder}
                        isCircularNavigation={true}
                        style={{ overflowY: 'scroll', height: '90%', padding: 0, marginTop: '10px' }}>
                        <Stack tokens={verticalGapStackTokens} style={{ margin: '0px 4px' }}>
                            {filterType[filter].listItems}
                        </Stack>{' '}
                    </FocusZone>
                </UIDialog>
            </div>
        </>
    );
}
