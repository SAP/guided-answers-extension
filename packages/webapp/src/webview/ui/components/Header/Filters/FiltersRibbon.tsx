import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import { AppState } from '../../../../types';
import { actions } from '../../../../state';
import { VscClose } from 'react-icons/vsc';
import './FiltersRibbon.scss';

/**
 *
 * @returns FiltersRibbon
 */
export function FiltersRibbon() {
    const appState = useSelector<AppState, AppState>((state) => state);
    const [selectedProductFilters, setSelectedProductFilters] = useState(appState.selectedProductFilters);
    const [selectedComponentFilters, setSelectedComponentFilters] = useState(appState.selectedProductFilters);
    const hasProductsFilter = selectedProductFilters.length > 0;
    const hasComponentsFilter = selectedComponentFilters.length > 0;
    const hasFilters = hasProductsFilter || hasComponentsFilter;
    const hasBothFilters = hasProductsFilter && hasComponentsFilter;
    const resetFilters = () => {
        actions.resetFilters();
        actions.searchTree({
            query: appState.query,
            filters: {
                product: [],
                component: []
            },
            paging: {
                responseSize: appState.pageSize,
                offset: 0
            }
        });
    };

    useEffect(() => {
        setSelectedProductFilters(
            appState.guidedAnswerTreeSearchResult.productFilters
                .filter((v: { PRODUCT: string }) => appState.selectedProductFilters.includes(v.PRODUCT))
                .map((v: { PRODUCT: any }) => v.PRODUCT)
        );
        setSelectedComponentFilters(
            appState.guidedAnswerTreeSearchResult.componentFilters
                .filter((v: { COMPONENT: string }) => appState.selectedComponentFilters.includes(v.COMPONENT))
                .map((v: { COMPONENT: any }) => v.COMPONENT)
        );
    }, [appState.selectedProductFilters, appState.selectedComponentFilters]);

    return (
        <>
            {hasFilters && (
                <div style={{ lineHeight: '18px' }}>
                    Searching in {hasProductsFilter ? 'Product' : ''}
                    {((hasProductsFilter && !hasComponentsFilter) || hasBothFilters) && <strong>&nbsp;</strong>}
                    <strong>
                        {selectedProductFilters && selectedProductFilters.map((pf: string) => pf).join(', ')}
                    </strong>
                    {hasBothFilters && <span>&nbsp; and &nbsp;</span>}
                    {hasComponentsFilter && !hasProductsFilter && <strong>&nbsp;</strong>}
                    {hasComponentsFilter ? 'Component' : ''}&nbsp;
                    <strong>
                        {selectedComponentFilters && selectedComponentFilters.map((cf: string) => cf).join(', ')}
                    </strong>
                    {hasBothFilters && <strong>&nbsp;</strong>}
                    <button className="clear-filters" onClick={resetFilters} title="Clear filters">
                        <VscClose className="clear-filters__content" />{' '}
                        <span className="clear-filters__content__text text-underline">Clear filters</span>
                    </button>
                </div>
            )}
        </>
    );
}
