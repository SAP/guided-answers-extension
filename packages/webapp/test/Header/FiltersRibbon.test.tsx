import React from 'react';
import { shallow } from 'enzyme';
import { FiltersRibbon } from '../../src/webview/ui/components/Header/Filters/FiltersRibbon';
import { actions } from '../../src/webview/state';

let state = {
    query: 'Fiori tools',
    selectedProductFilters: ['ProductFilter1, ProductFilter2'],
    selectedComponentFilters: ['ComponentFilter1', 'ComponentFilter2']
};

jest.mock('../../src/webview/state', () => {
    return {
        actions: {
            resetFilters: jest.fn(),
            searchTree: jest.fn()
        }
    };
});

jest.mock('react-redux', () => {
    const lib = jest.requireActual('react-redux');

    return {
        ...lib,
        useSelector: () => state
    };
});

describe('<FiltersRibbon />', () => {
    let wrapper: any;
    beforeEach(() => {
        wrapper = shallow(<FiltersRibbon />);
    });

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.unmount();
    });

    it('Should render a FiltersRibbon component', () => {
        const component = wrapper.html();
        expect(component).toMatchInlineSnapshot(
            `"<div style="line-height:18px">Searching in Product<strong> </strong><strong>ProductFilter1, ProductFilter2</strong><span>  and  </span>Component <strong>ProductFilter1, ProductFilter2</strong><strong> </strong><button class="clear-filters" title="Clear filters"><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16" class="clear-filters__content" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M8 8.707l3.646 3.647.708-.707L8.707 8l3.647-3.646-.707-.708L8 7.293 4.354 3.646l-.707.708L7.293 8l-3.646 3.646.707.708L8 8.707z"></path></svg> <span class="clear-filters__content__text text-underline">Clear filters</span></button></div>"`
        );

        wrapper.find('.clear-filters').simulate('click');
        expect(actions.resetFilters).toBeCalled();
        expect(actions.searchTree).toHaveBeenCalledWith({
            query: state.query,
            filters: {
                product: [],
                component: []
            },
            paging: {
                offset: 0,
                responseSize: undefined
            }
        });

        expect(actions.resetFilters).toHaveBeenCalledTimes(1);
        expect(actions.searchTree).toHaveBeenCalledTimes(1);
    });
});
