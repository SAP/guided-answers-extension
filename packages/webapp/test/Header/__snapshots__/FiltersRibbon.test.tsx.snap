// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FiltersRibbon /> Should render a FiltersRibbon component 1`] = `
<div>
  <div
    style="line-height: 18px; margin-top: 1px;"
  >
    Searching in 
    Product
    <strong>
       
    </strong>
    <strong>
      Product A
    </strong>
    <span>
        and  
    </span>
    Component
     
    <strong>
      comp-a
    </strong>
    <strong>
       
    </strong>
    <button
      class="clear-filters"
      id="clear-filters"
      title="Clear filters"
    >
      <i
        aria-hidden="true"
        class="ts-icon root-162"
        data-icon-name="Close"
      />
      <span>
        Clear filters
      </span>
    </button>
  </div>
</div>
`;
